### ISSUE TITLE
[Bug]: bundled flash_attn_v100 1.2.0 paged prefill kernels are nondeterministic — run-to-run |diff| up to 0.19 on identical inputs (official v1.2.1 wheel)

### ISSUE THREAD

**Issue by laris** [1CatAI/1Cat-vLLM-0.0.1#98]
### Environment
- Official wheel `1cat_vllm-1.2.1-cp312-cp312-linux_x86_64.whl`, installed exactly per the README (`--extra-index-url https://download.pytorch.org/whl/cu128`) → torch 2.10.0+cu128
- Tesla V100-SXM2-32GB (SM70), driver 580.126.09, Python 3.12
- Bundled `flash_attn_v100` reports version 1.2.0

### What we found
Kernel-level probe of the public paged APIs: **identical input tensors, 6 repeated calls**, measuring max run-to-run |diff| of the outputs (fp16):

| API | shape | max run-to-run \|diff\| |
|---|---|---|
| `flash_attn_prefill_paged` | block=256, L=300, HQ/HK=12/2, D=128 | **0.1917** |
| `flash_attn_prefill_paged` | block=256, L=512 (full pages) | **0.0999** |
| `flash_attn_prefill_paged` | block=16, L=300 | **0.1575** |
| `flash_attn_decode_paged` | same protocol | 0.000000 (bit-exact) |

Against an fp32 reference the prefill outputs intermittently show max_abs excursions of 0.05–0.11 (cos ≥ 0.99998) on any page size **including fully-aligned pages** — so this is not a paged-addressing problem. (We specifically audited page-straddle + garbage-tail + permuted-block-table addressing on both prefill and decode: all correct, cos=1.000000 on the good runs. Nice work on that, by the way — we went in looking for an addressing bug and found none.)

### Repro
```python
import torch
from flash_attn_v100 import flash_attn_prefill_paged
torch.manual_seed(0)
DEV, DT = "cuda", torch.float16
HQ, HK, D, L, block = 12, 2, 128, 300, 256
nb = (L + block - 1) // block
k = torch.randn(L, HK, D, device=DEV, dtype=DT); v = torch.randn(L, HK, D, device=DEV, dtype=DT)
kc = torch.full((nb, block, HK, D), 100.0, device=DEV, dtype=DT); vc = kc.clone()
for t in range(L):
    kc[t // block, t % block] = k[t]; vc[t // block, t % block] = v[t]
bt = torch.tensor([list(range(nb))], dtype=torch.int32, device=DEV)
sl = torch.tensor([L], dtype=torch.int32, device=DEV)
q = torch.randn(1, L, HQ, D, device=DEV, dtype=DT)
outs = [flash_attn_prefill_paged(q, kc, vc, bt, sl, causal=True).clone() for _ in range(6)]
print("max run-to-run |diff|:", max((a - b).abs().max().item() for a in outs for b in outs))
# observed on V100: up to ~0.19; expected for a deterministic kernel: 0.0
```

### Hypothesis
The extension exports `prefill_paged_splitkv_fwd` — the signature looks like a split-KV reduction race: non-fixed accumulation order (or a missing sync) when partial results are combined. Decode being bit-exact under the same protocol supports this.

### Impact
Silent run-to-run output variation at serving level (temperature-0 responses can differ across identical requests). To be clear: in our end-to-end serving tests (Qwen2.5-0.5B, Qwen2.5-14B-AWQ TP4, gemma-4-E2B/31B) we saw zero NaN/degenerate outputs and excellent performance — this is a determinism/precision-jitter report, not a correctness-catastrophe one.

### Ask
Is the split-KV prefill path known/expected to be nondeterministic? If yes, a doc note would help; a deterministic reduction (fixed combine order) would be better for temperature-0 reproducibility.

---
Found during an evaluation of V100 vLLM stacks (AI-assisted test harness; all results verified on physical hardware — DGX-1, 8× V100-SXM2-32GB). Full probe harness available on request.

**Comment by kezboard233** [comment]
We can confirm this at the serving level — your kernel-level finding (prefill nondeterministic, decode bit-exact) explains a reproducibility mystery we had been chasing for two days.

**Setup**: Qwen3.6-27B-AWQ, TP4 on 4x Tesla V100-SXM2-16GB, driver 535.216.03, v1.2.1, greedy (temperature=0), 480-token generation probes, output compared by SHA.

**Observation**: with identical config,

- a *narrative* probe (long free-form generation) flips between **two stable output SHAs across server restarts** — bistable, not random;
- *code* and *summarization* probes stay **bit-identical across every restart**;
- within one server lifetime, repeated runs of the same probe are SHA-stable — but that is likely masked by prefix caching (the second run's prefill is served from APC and never recomputed, and decode is bit-exact per your measurement).

This all lines up with `flash_attn_prefill_paged` being run-to-run nondeterministic while decode is exact: a fresh prefill after restart lands on a slightly different numeric state; a long low-margin narrative generation forks at some early low-confidence token, then greedy decode deterministically follows whichever branch was entered. High-margin outputs (code, extractive summaries) absorb the perturbation.

Practical impact worth noting in triage: SHA-anchored regression testing on SM70 currently requires either cache-warm comparisons or accepting multiple anchors for prefill-sensitive prompts. A deterministic (even if slower, opt-in) prefill reduction order would be very welcome.

**Comment by yangzhuxinyzx** [comment]
Root cause identified: this reproduction uses the generic D=128 `prefill_paged_fwd` kernel, not split-KV (the latter is D=256-only). In the v1.2.1 layout its FP32 WMMA accumulator shared buffer started at byte offset 80,144, which is only 16-byte aligned. CUDA WMMA `load_matrix_sync` / `store_matrix_sync` require a 256-bit-aligned matrix pointer, so this is undefined behavior.

Draft fix: https://github.com/1CatAI/1Cat-vLLM/pull/123

The patch explicitly aligns the accumulator and dynamic shared base, adds a compile-time assertion, and adds a six-replay D=128 regression for both the extension entrypoint and Python wrapper. Offline SM70 compilation and extension import pass. The remaining merge gate is running that regression and a temperature-0 serving reproducibility probe on a physical V100; the local host currently has CUDA driver error 804.
